### PR TITLE
Get run last metrics optimization

### PR DESCRIPTION
**Pull request by t-henri**
## What changes are proposed in this pull request?
 
To improve the speed of the UI when it has to load a lot of runs with lots of time series metrics, we changed the way we get the last recorded metric for each run. We removed the python loop that loops over all metrics by a SQL query that directly gets the last recorded metrics.
Following this issue: https://github.com/mlflow/mlflow/issues/1571

I compared the execution time before and after this change by directly calling the search_runs method of SqlAlchemyStore and timing it with %%prun in a jupyter notebook.
In our case, it went from 11.3 seconds to 3.7 seconds.
We have 617 runs with 202967 metrics in total.
 
## How is this patch tested?
 
Using the tests already present in the repository (we have not changed the current functionalities)
 
### Is this a user-facing change? 

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)
 
### What component(s) does this PR affect?
 
- [ ] UI
- [ ] CLI 
- [ ] API 
- [ ] REST-API 
- [ ] Examples 
- [ ] Docs
- [X] Tracking
- [ ] Projects 
- [ ] Artifacts 
- [ ] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [ ] Java
- [X] Python

### How should the PR be classified in the release notes? Choose one:
 
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/store/dbmodels/models.py [hunk 1]
```diff
@@ -187,6 +192,33 @@ def to_mlflow_entity(self):
...
+            .subquery('metrics_with_max_timestamp')
+        metrics_with_max_value = session \
+            .query(SqlMetric.run_uuid, SqlMetric.key, SqlMetric.step, SqlMetric.timestamp,
+                   func.max(SqlMetric.value).label('value'), SqlMetric.is_nan) \
+            .join(metrics_with_max_timestamp,
+                  and_(SqlMetric.timestamp == metrics_with_max_timestamp.c.timestamp,
+                       SqlMetric.run_uuid == metrics_with_max_timestamp.c.run_uuid,
+                       SqlMetric.key == metrics_with_max_timestamp.c.key,
+                       SqlMetric.step == metrics_with_max_timestamp.c.step)) \
+            .filter(SqlMetric.run_uuid == self.run_uuid) \
```
Comment (dbczumar) [inline mlflow/store/dbmodels/models.py]: I think we should filter before each join to further improve the performance.

**Review by dbczumar — COMMENTED** [review]
@t-henri Thank you for this contribution! I tried running this against MySQL and encountered an exception that appears to be caused by the fact that `SqlMetric.is_nan` is present in the `SELECT` statement but absent from the `GROUP BY`:

```
Traceback (most recent call last):
  File "logmetrics.py", line 10, in <module>
    with mlflow.start_run() as active_run:
  File "/Users/czumar/mlflow/mlflow/tracking/fluent.py", line 143, in start_run
    tags=tags
  File "/Users/czumar/mlflow/mlflow/tracking/client.py", line 88, in create_run
    tags=[RunTag(key, value) for (key, value) in iteritems(tags)]
  File "/Users/czumar/mlflow/mlflow/store/sqlalchemy_store.py", line 348, in create_run
    return run.to_mlflow_entity(session)
  File "/Users/czumar/anaconda2/lib/python2.7/contextlib.py", line 35, in __exit__
    self.gen.throw(type, value, traceback)
  File "/Users/czumar/mlflow/mlflow/store/sqlalchemy_store.py", line 154, in make_managed_session
    raise MlflowException(message=e, error_code=INTERNAL_ERROR)
mlflow.exceptions.MlflowException: (MySQLdb._exceptions.OperationalError) (1055, "Expression #6 of SELECT list is not in GROUP BY clause and contains nonaggregated column 'opt.metrics.is_nan' which is not functionally dependent on columns in GROUP BY clause; this is incompatible with sql_mode=only_full_group_by")
[SQL: SELECT metrics.run_uuid AS metrics_run_uuid, metrics.`key` AS metrics_key, metrics.step AS metrics_step, metrics.timestamp AS metrics_timestamp, max(metrics.value) AS value, metrics.is_nan AS metrics_is_nan
FROM metrics INNER JOIN (SELECT metrics.run_uuid AS run_uuid, metrics.`key` AS `key`, metrics.step AS step, max(metrics.timestamp) AS timestamp
FROM metrics INNER JOIN (SELECT metrics.run_uuid AS run_uuid, metrics.`key` AS `key`, max(metrics.step) AS step
FROM metrics
WHERE metrics.run_uuid = %s GROUP BY metrics.`key`, metrics.run_uuid) AS metrics_with_max_step ON metrics.step = metrics_with_max_step.step AND metrics.run_uuid = metrics_with_max_step.run_uuid AND metrics.`key` = metrics_with_max_step.`key`
WHERE metrics.run_uuid = %s GROUP BY metrics.`key`, metrics.run_uuid) AS metrics_with_max_timestamp ON metrics.timestamp = metrics_with_max_timestamp.timestamp AND metrics.run_uuid = metrics_with_max_timestamp.run_uuid AND metrics.`key` = metrics_with_max_timestamp.`key` AND metrics.step = metrics_with_max_timestamp.step
WHERE metrics.run_uuid = %s GROUP BY metrics.`key`, metrics.run_uuid]
[parameters: ('56bfde81bab040ad999233ea97e68797', '56bfde81bab040ad999233ea97e68797', '56bfde81bab040ad999233ea97e68797')]
(Background on this error at: http://sqlalche.me/e/e3q8)
```

Additionally, I left a comment about filtering before each join. I'll submit a PR against your branch to address these issues tomorrow morning, but feel free to address them before then!

**Comment by t-henri** [comment]
Hello,

Thanks for your feedback, I'm willing to resolve these issues, but could you indicate me how to reproduce the bug you faced ? I didn't encounter it.

**Comment by dbczumar** [comment]
@t-henri Sorry about that! I encountered the issue when using MySQL version `8.0.15` with `only_full_group_by` mode enabled (see https://dev.mysql.com/doc/refman/5.7/en/sql-mode.html#sqlmode_only_full_group_by).

**Comment by dbczumar** [comment]
@t-henri I filed https://github.com/criteo-forks/mlflow/pull/5 with the proposed changes. Please take a look and let me know what you think.

**Comment by t-henri** [comment]
@dbczumar I took a look at your pull request and it looks good to me.
I tried it and we have extra performance improvement with the run_uuid filter in the joins, and the only_full_group_by problem is gone, so I merged it.

Let me know if you have any other remark.

**Review by dbczumar — APPROVED** [review]
@t-henri I pushed another commit that modifies each query/subquery to explicitly group by columns that are not aggregated within the `SELECT` statement. Postgres seems to require this specification, as I encountered the following error when attempting to run logging code against Postgres:

```
mlflow.exceptions.MlflowException: (psycopg2.errors.GroupingError) column "metrics.step" must appear in the GROUP BY clause or be used in an aggregate function
LINE 2: ...metrics.run_uuid AS run_uuid, metrics.key AS key, metrics.st...
                                                             ^

[SQL: SELECT metrics.run_uuid AS metrics_run_uuid, metrics.key AS metrics_key, metrics.step AS metrics_step, metrics.timestamp AS metrics_timestamp, max(metrics.value) AS value, metrics.is_nan AS metrics_is_nan
FROM metrics JOIN (SELECT metrics.run_uuid AS run_uuid, metrics.key AS key, metrics.step AS step, max(metrics.timestamp) AS timestamp
FROM metrics JOIN (SELECT metrics.run_uuid AS run_uuid, metrics.key AS key, max(metrics.step) AS step
FROM metrics
WHERE metrics.run_uuid = %(run_uuid_1)s GROUP BY metrics.key, metrics.run_uuid) AS metrics_with_max_step ON metrics.step = metrics_with_max_step.step AND metrics.run_uuid = metrics_with_max_step.run_uuid AND metrics.key = metrics_with_max_step.key
WHERE metrics.run_uuid = %(run_uuid_2)s GROUP BY metrics.key, metrics.run_uuid) AS metrics_with_max_timestamp ON metrics.timestamp = metrics_with_max_timestamp.timestamp AND metrics.run_uuid = metrics_with_max_timestamp.run_uuid AND metrics.key = metrics_with_max_timestamp.key AND metrics.step = metrics_with_max_timestamp.step
WHERE metrics.run_uuid = %(run_uuid_3)s GROUP BY metrics.key, metrics.run_uuid, metrics.is_nan]
```

Logging code:
```
import mlflow
from mlflow.tracking.client import MlflowClient

mlflow.set_tracking_uri("postgresql://root:password@localhost:5432/mydb")

client = MlflowClient()

for iteration in range(10):
    with mlflow.start_run() as active_run:
        run_id = active_run.info.run_id

        for i in range(20):
            for j in range(7):
                step = i % 10
                timestamp = j % 3

                value = (step + timestamp) % 4

                client.log_metric(run_id=run_id, key="mymetric", value=value, step=step, timestamp=timestamp)
```

After this change, the PR seems to be in a good state. Please let me know if you have any questions or feedback; I intend to merge tomorrow morning if everything looks good from your end!

**Comment by t-henri** [comment]
@dbczumar Looks good to me ! I just pushed a minor linter fix.

**Pull request merged**